### PR TITLE
Merge 0.11.x:  mimeencoding, flyweight code RF

### PR DESCRIPTION
**Pull request by yarikoptic**
Conflicts were not fun.  but seems to test selected tests I ran locally, so I guess I didn't ruin it.

### CODE REVIEW

**Comment by kyleam** [comment]
> Conflicts were not fun.

Thanks for handling it.  I bet I'll relive your fun as I review it.

**Comment by yarikoptic** [comment]
well -- the ultimate diff here is quite easier to digest than 3way diff of merge conflicts which were there ;)
FWIW: not sure what's up with coverage

File: datalad/distribution/dataset.py [hunk 1]
```diff
@@ -121,29 +100,25 @@ def _flyweight_id_from_args(cls, *args, **kwargs):
...
-        kwargs['path'] = path_
-        return path_, args, kwargs
+        return op.normpath(path)
 
     def _flyweight_invalid(self):
         """Invalidation of Flyweight instance
 
-        Dataset doesn't need to be invalidated during its lifetime at all. Instead the underlying *Repo instances are.
+        Dataset doesn't need to be invalidated during its lifetime at all.
+        Instead the underlying *Repo instances are.
```
Comment (kyleam) [inline datalad/distribution/dataset.py]: When dealing with merge conflicts, please avoid this sort of unrelated formatting.  It introduces changes that aren't shared by any side, which are typically the changes of most interest, but here they're just noise.

File: datalad/distribution/dataset.py [hunk 2]
```diff
@@ -36,9 +36,10 @@
     InvalidGitRepositoryError,
     NoSuchPathError
 )
-from datalad.support.repo import Flyweight
+from datalad.support.repo import PathBasedFlyweight
 from datalad.support.network import RI
 from datalad.support.exceptions import InvalidAnnexRepositoryError
+from datalad.support import path as op
```
Comment (kyleam) [inline datalad/distribution/dataset.py]: This is introducing a duplicate import of `op`.

File: datalad/interface/tests/test_run_procedure.py [hunk 3]
```diff
@@ -304,17 +305,52 @@ def test_quoting(path):
...
+    'emptyline': '\n',  # libmagic: "binary" "application/octet-stream"
+})
+def test_text2git(path):
+    # Test if files being correctly annexed in a ds configured with text2git.
+    TEXT_FILES = ('JSON', 'YAML', 'MARKDOWN', 'empty')
+    BINARY_FILES = ('0blob', 'emptyline')
+
     ds = Dataset(path).create(force=True)
     ds.run_procedure('cfg_text2git')
+
```
Comment (kyleam) [inline datalad/interface/tests/test_run_procedure.py]: This space is another change that isn't from either side.

File: datalad/interface/tests/test_run_procedure.py [hunk 4]
```diff
@@ -304,17 +305,52 @@ def test_quoting(path):
...
-    ds.save(message="add empty file")
-    # check that it's not annexed
-    assert_false(ds.repo.is_under_annex("emptyfile"))
+
+    # check that text files are not annexed
+    for f in TEXT_FILES:
+        assert_false(ds.repo.is_under_annex(f))
+    # and trivial binaries - annexed
+    for f in BINARY_FILES:
+        assert_true(ds.repo.is_under_annex(f))
```
Comment (kyleam) [inline datalad/interface/tests/test_run_procedure.py]: Hmm, da27e86c4 (part of gh-3956) dropped the `assert_false(ds.repo.is_under_annex("emptyfile"))` above for these `ok_file_under_git` calls.  Why aren't you keeping that change here?

File: datalad/distribution/dataset.py (hunk 1 above)
Comment (yarikoptic) [reply datalad/distribution/dataset.py]: not 100% certain I did it consciously ... generally I am trying to avoid such changes while working on merges.  do you think it is worth rewriting for this one?

File: datalad/interface/tests/test_run_procedure.py (hunk 4 above)
Comment (yarikoptic) [reply datalad/interface/tests/test_run_procedure.py]: this whole test was pretty much rewritten from being a test for only an empty file, to a test which tests various types (text and "binary") of files -- from `test_text2git_empty` to `test_text2git`. So `diff` here could be a bit misleading - I would not strive to minimize diff here, nor re-adding `assert_false(ds.repo.is_under_annex("emptyfile"))` would add any additional testing

File: datalad/interface/tests/test_run_procedure.py (hunk 3 above)
Comment (yarikoptic) [reply datalad/interface/tests/test_run_procedure.py]: see comment below

File: datalad/distribution/dataset.py (hunk 1 above)
Comment (kyleam) [reply datalad/distribution/dataset.py]: That'd be my personal preference, but given what I perceive to be the preference of this project in general, I'd say "no, not if this were the only issue".  But if you agree that the `ok_file_under_git` point below should be addressed anyway, then I think it'd be good to resolve these minor issues in the same rewrite.

File: datalad/distribution/dataset.py (hunk 2 above)
Comment (yarikoptic) [reply datalad/distribution/dataset.py]: rright -- removing `import os.path as op` above. pushed 37e9a9e05599b271cc5743f3cb0155babcf9c197

File: datalad/interface/tests/test_run_procedure.py (hunk 4 above)
Comment (kyleam) [reply datalad/interface/tests/test_run_procedure.py]: Let me try again.  Here is the resolution that you pushed:

```
+     ds.save(path=TEXT_FILES + BINARY_FILES, message="added all files")
 -    ok_clean_git(ds.path)
 +    assert_repo_status(ds.path)
-     # create an empty file, no extension
-     open(op.join(path, 'emptyfile'), 'a').close()
-     ds.save(message="add empty file")
-     # check that it's not annexed
-     assert_false(ds.repo.is_under_annex("emptyfile"))
+ 
+     # check that text files are not annexed
+     for f in TEXT_FILES:
 -        ok_file_under_git(path, f, annexed=False)
++        assert_false(ds.repo.is_under_annex(f))
+     # and trivial binaries - annexed
+     for f in BINARY_FILES:
 -        ok_file_under_git(path, f, annexed=True)
++        assert_true(ds.repo.is_under_annex(f))
```

Note that you're introducing two changes (the `++` liines) that come from neither side.  Unless there is a master-specific reason for why doing that, you should not.

File: datalad/interface/tests/test_run_procedure.py (hunk 3 above)
Comment (kyleam) [reply datalad/interface/tests/test_run_procedure.py]: I don't see how the comment below is relevant.  Even if the test was rewritten on 0.11.x's side (specifically da27e86), the resolution shouldn't introduce a space that wasn't in 0.11.x or in master.

File: datalad/interface/tests/test_run_procedure.py (hunk 4 above)
Comment (yarikoptic) [reply datalad/interface/tests/test_run_procedure.py]: hm, IIRC I was just replacing "0.11.x" style (`ok_file_under_git(path, f, annexed=False)`) from one side to "master style" (`assert_false(ds.repo.is_under_annex(f))`).  What should have I done differently?

File: datalad/interface/tests/test_run_procedure.py (hunk 3 above)
Comment (yarikoptic) [reply datalad/interface/tests/test_run_procedure.py]: oh, gotcha. True!  Let's see if we agree on the below and then I could rewrite that merge to make it clearer (to avoid possible future conflicts)... or you are most welcome to do so and force push here!

Comment (kyleam) [reply datalad/interface/tests/test_run_procedure.py]: > you are most welcome to do so and force push here!

OK, I'll do that.

File: datalad/interface/tests/test_run_procedure.py (hunk 4 above)
Comment (kyleam) [reply datalad/interface/tests/test_run_procedure.py]: > I was just replacing "0.11.x" style (`ok_file_under_git(path, f, annexed=False)`) from one side to "master style" (`assert_false(ds.repo.is_under_annex(f))`)

Thanks, knowing your intention here helps.

Given that the `assert_false(ds.repo.is_under_annex(f))` line was shared with 0.11.x, and you just did the reverse substitution in 0.11.x's da27e86, in my view the commit message should explain why we should prefer one form on master and the other on 0.11.x, ideally with a pointer to where we've discussed `ok_file_under_git` being problematic.  The main thing I can find is gh-2961, but it looks like that was resolved by your gh-2972.

Even more ideally, if `ok_file_under_git` is still an issue and should be avoided on master, we'd also avoid it on 0.11.x, because (1) it seems likely that the same issues would be relevant for 0.11.x and (2) it introduces needless differences when the master variant will work fine on 0.11.x.

I don't think it is worth a PR to 0.11.x at this point, so I'll go with adding something to the commit message to the merge that I force-push here.

Comment (yarikoptic) [reply datalad/interface/tests/test_run_procedure.py]: thanks for the analysis! I just saw difference, so assumed that it would be the right way to do in master without checking compatibility with 0.11.x .  Please do as you see it fit! Thanks in advance!

**Comment by kyleam** [comment]
AppVeyor failure is the familiar "Invoke-WebRequest : Unable to connect" failure.  Will merge.
